external/java/src/main/java/ch/epfl/dedis/byzcoin/Proof.java

```diff
@@ -112,64 +159,183 @@ public StateChangeBody getValues() {
-     * @return true if this is a matching byzcoin proof for a contract of the given contract.
+     * Check whether the key on the leaf exists.
+     *
+     * @return true if the proof has the key/value pair stored on the leaf, false if it
+     * is a proof of absence or an error has occured.
      */
-    public boolean isContract(String expected){
-        if (!isByzCoinProof()){
+    public boolean matches() {
+        if (proof.getLeaf().getKey().isEmpty()) {
             return false;
         }
-        if (!getContractID().equals(expected)) {
+        try {
+            return this.exists(proof.getLeaf().getKey().toByteArray());
+        } catch (CothorityCryptoException e) {
             return false;
         }
-        return true;
     }
 
     /**
-     * Checks if the proof is valid and of type expected.
+     * Check whether the key exists in the proof.
      *
-     * @param expected the expected contractId
-     * @param id the Byzcoin id to verify the proof against
-     * @return true if the proof is correct with regard to that Byzcoin id and the contract is of the expected type.
-     * @throws CothorityException if something goes wrong
+     * @param key is the value that we want to proof whether it exists or is absent.
+     * @return true if it's an existence proof.
+     * @throws CothorityCryptoException if an unexpected error occurs, an absence proof does not throw an exception.
      */
-    public boolean isContract(String expected, SkipblockId id) throws CothorityException{
-        if (!verify(id)){
-            return false;
+    public boolean exists(byte[] key) throws CothorityCryptoException {
+        if (key == null) {
+            throw new CothorityCryptoException("key is nil");
+        }
+
+        if (this.proof.getInteriorsCount() == 0) {
+            throw new CothorityCryptoException("no interior nodes");
+        }
+
+        Boolean[] bits = binSlice(key);
+        byte[] expectedHash = hashInterior(this.proof.getInteriors(0));
+
+        int i;
+        for (i = 0; i < this.proof.getInteriorsCount(); i++) {
+            if (!Arrays.equals(expectedHash, hashInterior(this.proof.getInteriors(i)))) {
+                return false;
+            }
+            if (bits[i]) {
+                expectedHash = this.proof.getInteriors(i).getLeft().toByteArray();
+            } else {
+                expectedHash = this.proof.getInteriors(i).getRight().toByteArray();
+            }
         }
-        if (!isContract(expected)){
+        // we use i below instead of i+1 (like the go code) because the final i is one more than the i used in the final iteration
+        if (Arrays.equals(expectedHash, hashLeaf(this.proof.getLeaf(), this.proof.getNonce().toByteArray()))) {
+            if (!Arrays.equals(Arrays.copyOf(bits, i), this.proof.getLeaf().getPrefixList().toArray(new Boolean[0]))) {
+                throw new CothorityCryptoException("invalid prefix in leaf node");
```

Comment:
Is it important to test this? It seems to me like we are only testing the happy path through here. What if a nasty conode is sending out Proofs to clients that are invalid? Are we sure they will catch it?

I wonder if it would be enough to take a correct proof and then fiddle with some of it's internals, then call verify() again on it, in order to check for the expected exceptions?

Reply:
moving this back to WIP while I think about how we can test something like this

Reply:
Added better tests - https://github.com/dedis/cothority/pull/1654/commits/6d8d31d9ee8893b59163edfab865ade938da290b